Marlin/src/HAL/HAL_STM32/persistent_store_flash.cpp

```diff
@@ -0,0 +1,256 @@
+ * on 2 of these pages. Each write, we'd use 2 different pages from a pool of pages until we are done.
+ */
+#if ENABLED(FLASH_EEPROM_LEVELING) && (defined(STM32F407xx) || defined(STM32F446xx))
+
+  #include "stm32_def.h"
+
+  #define DEBUG_OUT ENABLED(FLASH_DEBUG)
+  #include "src/core/debug_out.h"
+
+  #ifndef FLASH_SECTOR
+    #define FLASH_SECTOR          (FLASH_SECTOR_TOTAL - 1)
+  #endif
+  #ifndef FLASH_UNIT_SIZE
+    #define FLASH_UNIT_SIZE       0x20000 // 128kB
+  #endif
+  #ifndef FLASH_ADDRESS_START
+    #define FLASH_ADDRESS_START   0x080E0000
+  #endif
+  #ifndef EEPROM_SIZE
+    #define EEPROM_SIZE           0x1000    // 4kB, we could bump this up if we want to go crazy with ubl meshes
+  #endif
+
+  #define FLASH_ADDRESS_END       (FLASH_ADDRESS_START + FLASH_UNIT_SIZE)
+  #define EEPROM_SLOTS            (FLASH_UNIT_SIZE/EEPROM_SIZE)
+  #define SLOT_ADDRESS(slot)      (FLASH_ADDRESS_START + (slot * EEPROM_SIZE))
+
+  //#define DISABLE_IRQ()         prim = __get_PRIMASK();__disable_irq()
+  //#define ENABLE_IRQ()          if (!prim) __enable_irq()
+
+  #define UNLOCK_FLASH()          if (!flash_unlocked) { \
+                                    HAL_FLASH_Unlock(); \
+                                    __HAL_FLASH_CLEAR_FLAG(FLASH_FLAG_EOP | FLASH_FLAG_OPERR | FLASH_FLAG_WRPERR | \
+                                                           FLASH_FLAG_PGAERR | FLASH_FLAG_PGPERR | FLASH_FLAG_PGSERR); \
+                                    flash_unlocked = true; \
+                                  }
+  #define LOCK_FLASH()            if (flash_unlocked) { HAL_FLASH_Lock(); flash_unlocked = false; }
+
+  #define EMPTY_UINT32            ((uint32_t)-1)
+  #define EMPTY_UINT8             ((uint8_t)-1)
+
+  static uint8_t ram_eeprom[EEPROM_SIZE] __attribute__((aligned(4))) = {0};
+  static int current_slot = 0;
+#else
+  #undef FLASH_EEPROM_LEVELING
+#endif
+
+static bool eeprom_data_written = false;
+
+bool PersistentStore::access_start() {
+
+  #ifdef FLASH_EEPROM_LEVELING
+
+    static_assert(true == IS_FLASH_SECTOR(FLASH_SECTOR), "FLASH_SECTOR is invalid");
+    static_assert(0 == EEPROM_SIZE % 4, "EEPROM_SIZE must be a multiple of 4"); // Ensure copying as uint32_t is safe
+    static_assert(0 == (FLASH_UNIT_SIZE & (FLASH_UNIT_SIZE - 1)), "FLASH_UNIT_SIZE should be a power of 2, please check your chips spec sheet");
+    static_assert(0 == FLASH_UNIT_SIZE % EEPROM_SIZE, "EEPROM_SIZE must divide evenly into your FLASH_UNIT_SIZE");
+
+    current_slot = -1;
+
+    // Look to see if we have any data stored yet, and where
+    uint32_t address = FLASH_ADDRESS_START;
+    while (address < FLASH_ADDRESS_END) {
+      uint32_t address_value = (*(__IO uint32_t*)address);
+      if (address_value != EMPTY_UINT32) {
+        current_slot = (address - FLASH_ADDRESS_START) / EEPROM_SIZE;
+        break;
+      }
+      address += sizeof(uint32_t);
+    }
```

Comment:
I haven't looked back at what the LPC code does, but do we need to check every uint32_t value to find the first unused one? Could one single location in each slot be checked? I suppose it is possible to have only UBL meshes saved, without ever saving EEPROM configuration, so that might be tricked than it sounds.

Reply:
Ideally the last word written (in time, not necessary location) in each slot would be a "valid" indicator, that could be used to ensure a slot is only used if it was written completely. That would allow ignoring a corrupt slot, where power was lost while writing the flash. That of course could not protect you when the entire sector is erased. More than one sector would be needed to provide fault tolerance in that case.

It probably doesn't make sense to tie up this feature for that, unless it could already be easily implemented without requiring EEPROM layout changes that would result in issues elsewhere.

Reply:
>  ignoring a corrupt slot

If you chose to implement this it would also add complexity where you'd have to skip corrupt slots when choosing your destination, rather than simply rolling to the next slot.

Reply:
> I haven't looked back at what the LPC code does, but do we need to check every uint32_t value to find the first unused one? Could one single location in each slot be checked? I suppose it is possible to have only UBL meshes saved, without ever saving EEPROM configuration, so that might be tricked than it sounds.

The configuration in the 4k buffer is offset by 100, with the print counter information stored before it. We could, probably look around offset 100 where it stores the checksum.

The LPC code calls a blanksector function and doesn't have that type of loop inside the HAL. I can dig into that function and check.

Reply:
One thing that'll make things tricky is if we adapt this code across the other chips. The sector concept of the F4 chips, where they have 8 to 12 "blocks" between 16k and 128k is easy to work with because we can use one "sector" and subdivide.

The F103 chips, which currently uses a completely different HAL for now, uses a "page" concept. Only difference is the number and size. Instead of subdividing what I'll just call a flash "unit" into multiple slots, we'd have to subdivide a slot into multiple units (or pages). 

The datasheets show chips with pages of 1, 2, and 4k. It could get convoluted quickly.

Would we want to get more variations of these STM32 chips supported with this type of wear leveling and then add the corruption check add it first?